packages/shopping-cart/src/shopping-cart-manager.ts

```diff
@@ -0,0 +1,149 @@
+import debugFactory from 'debug';
+import {
+	getShoppingCartManagerState,
+	createSubscriptionManager,
+	createActionPromisesManager,
+	noopManager,
+} from './managers';
+import { createActions } from './shopping-cart-actions';
+import {
+	getInitialShoppingCartState,
+	isStatePendingUpdateOrQueuedAction,
+	reducerWithQueue as shoppingCartReducer,
+} from './shopping-cart-reducer';
+import { createTakeActionsBasedOnState } from './state-based-actions';
+import { createCartSyncManager } from './sync';
+import type {
+	GetCart,
+	SetCart,
+	ShoppingCartManagerClient,
+	ShoppingCartManagerState,
+	ShoppingCartManager,
+	ShoppingCartAction,
+	ResponseCart,
+	ShoppingCartReducerDispatch,
+	DispatchAndWaitForValid,
+	ActionPromises,
+	ShoppingCartState,
+} from './types';
+
+const debug = debugFactory( 'shopping-cart:shopping-cart-manager' );
+
+function createDispatchAndWaitForValid(
+	dispatch: ShoppingCartReducerDispatch,
+	actionPromises: ActionPromises
+): DispatchAndWaitForValid {
+	return ( action ) => {
+		return new Promise< ResponseCart >( ( resolve ) => {
+			actionPromises.add( resolve );
+			dispatch( action );
+		} );
+	};
+}
+
+function createShoppingCartManager(
+	cartKey: string,
+	getCart: GetCart,
+	setCart: SetCart
+): ShoppingCartManager {
+	let state = getInitialShoppingCartState();
+
+	const subscriptionManager = createSubscriptionManager( cartKey );
+
+	const syncManager = createCartSyncManager( cartKey, getCart, setCart );
+	const actionPromises = createActionPromisesManager();
+	const takeActionsBasedOnState = createTakeActionsBasedOnState( syncManager );
+
+	let deferredStateCheck: ReturnType< typeof setTimeout >;
+	const dispatch = ( action: ShoppingCartAction ) => {
+		debug( `dispatching action for cartKey ${ cartKey }`, action.type );
+		const newState = shoppingCartReducer( state, action );
+		if ( newState === state ) {
+			if ( ! isStatePendingUpdateOrQueuedAction( state ) ) {
+				debug( 'state did not change; resolving action promises' );
+				actionPromises.resolve( state.responseCart );
+			}
+			return;
+		}
+		state = newState;
+
+		// We defer the state based actions so that multiple cart changes can be
+		// batched together during the same run of the event loop.
+		if ( deferredStateCheck ) {
+			clearTimeout( deferredStateCheck );
+		}
+		deferredStateCheck = setTimeout( () => {
+			takeActionsBasedOnState( state, dispatch );
+		} );
```

Comment:
An idea: move the `setTimeout` logic to the `takeActionsBasedOnState` helper, so that it's encapsulated there and `dispatch` can simply call a single function.

The `fetchInitialCart` function doesn't mind if the sync is started synchronously, or deferred.